Preserve cause of exception when getting/setting attribute in DeviceProxy

woutdenolf opened the pull request:
Closes #364

woutdenolf committed:
[deviceproxy] preserve cause of exception when getting/setting attribute

woutdenolf commented:
I guess we still need to support python 2?

```python
from six import raise_from
```

ajoubertza commented:
> I guess we still need to support python 2?

Yes, please!

ajoubertza commented on tango/device_proxy.py:
```diff
@@ -293,38 +294,42 @@ def __DeviceProxy__getattr(self, name):
-    except:
-        pass
+    except Exception as e:
+        if cause is None:
+            cause = e
```
There would be multiple exceptions in this method if the device is off/unreachable, so keeping the first exception is a good approach 👍

ajoubertza commented on tango/device_proxy.py:
```diff
@@ -334,29 +339,36 @@ def __DeviceProxy__setattr(self, name, value):
 
-    return super(DeviceProxy, self).__setattr__(name, value)
+    try:
+        return super(DeviceProxy, self).__setattr__(name, value)
+    except Exception as e:
+        raise e from cause
```
This case is interesting.  If we get here then `name` isn't a known Tango device command/attribute/pipe.  We end up writing to either an existing Python class instance attribute, or adding a new one, if it doesn't exist yet.

I'm not sure that `__setattr__` can even raise an exception?  If it did, the `cause` may or may not be relevant.

A couple of cases could occur for the example of an unreachable device:
1. We haven't yet cached the lists of commands, attributes and pipes.  
2. We have cached the lists of commands, attributes and pipes prior to this call. 

Then we get these four scenarios, depending on if the Tango command/attribute/pipe either:

- Exists on the device:
  - For case 1:  we'll write to a local class instance attribute and the user won't even get an error.  This behaviour is unexpected, but not a new problem.
  - For case 2: we would have failed because it is a command, or we tried to call `__set_attribute_value` or `write_pipe` earlier in this method, so the user will get whatever error those methods throw.  That's fine.
- Doesn't exist on the device:
  - For case 1:  we'll write to a local class instance attribute and the user won't get an error.  This could be the desired behaviour, if the user was storing some other info on an instance of a `DeviceProxy`.  Or, it could be undesired, if the user thought they were setting a Tango attribute/pipe but had the wrong name.  Not much we can do about that. 
  - For case 2: same as case 1.  There won't be an error, and we're not sure if the user expected one of not.

In conclusion, the change in the PR is fine.  It has just made us aware of some nuances in how setting attributes on a `DeviceProxy` behaves.

ajoubertza commented on tango/device_proxy.py:
```diff
@@ -334,29 +339,36 @@ def __DeviceProxy__setattr(self, name, value):
+        return super(DeviceProxy, self).__setattr__(name, value)
+    except Exception as e:
+        raise e from cause
+
```
We don't need this blank line.  Just 2 between functions according to PEP8.  I'm not sure why Codacy didn't complain.

ajoubertza reviewed (changes requested):
The approach looks good, just need the Python 2 support.

woutdenolf committed:
[deviceproxy] make `raise ... from ...` py2 compatible

woutdenolf committed:
[deviceproxy] remove blank line

ajoubertza committed:
Avoid Py3-only version of pytest-xdist on Py2

For some reason the CI tests for Python 2.7 are attempting to install a version of pytest-xdist that no longer supports Python 2.7.  Explicitly limit the version to avoid this.

ajoubertza commented on tango/device_proxy.py:
```diff
@@ -334,29 +339,35 @@ def __DeviceProxy__setattr(self, name, value):
+            cause = e
 
     if name_l in self.__get_cmd_cache():
         raise TypeError('Cannot set the value of a command')
```
Shouldn't this line include `cause`, similar to line 332?

woutdenolf replied:
Good catch. Thanks!

ajoubertza reviewed (commented):
Thanks for the update.  I have one query.

(And I updated the test requirements due to the last CI failure)

woutdenolf committed:
[deviceproxy] missing `raise ... from ...` in attribute setter

ajoubertza reviewed (approved):
Looks good.  Thanks for the contribution!
